File: java/euler19b.java
class Euler19b {

	static int monthDays[] = {
		31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31
	};

	private static boolean isLeapYear(int year) {
		return ((year % 4 == 0) && (year % 100 != 0))
			|| (year % 400 == 0);
	}

	public static void main(String[] args) {
		int count = 0;
		
		// 01/jan/1900 is monday = 1, so
		// 31/dec/1899 is sunday = 0
		int baseDay = 1;

		// advance to 1/jan/1901
		baseDay += isLeapYear(1900) ? 366 : 365;
		
		for (int year = 1901; year <= 2000; year++) {
			for (int month = 0; month < 12; month++) {
				if (baseDay % 7 == 0) {
					count++;
				}
				baseDay += monthDays[month];
				if (isLeapYear(year) && month == 1) {
					baseDay++;
				}
			}
		}
		
		System.out.println(count);
	}
}

File: haskell/euler19b.hs
{-
Counting Sundays
Problem 19
You are given the following information, but you may prefer to do some research for yourself.

1 Jan 1900 was a Monday.
Thirty days has September,
April, June and November.
All the rest have thirty-one,
Saving February alone,
Which has twenty-eight, rain or shine.
And on leap years, twenty-nine.
A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.

How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?
-}

-- Using library functions (my implementation is on euler19.hs)

import Data.Time.Calendar
import Data.Time.Calendar.OrdinalDate

initial = fromGregorian 1901 1 1
final = fromGregorian 2000 12 31
period = [initial..final]

isSundayFirst :: Day -> Bool
isSundayFirst date = 
	let (year, month, day) = toGregorian date
	    (week, weekday) = mondayStartWeek date
	 in day == 1 && weekday == 7

sundayFirsts = filter isSundayFirst period

euler19 = length sundayFirsts

-- A more direct approach

euler19' = sum [ 1 | m <- [1..12], y <- [1901..2000], snd (mondayStartWeek (fromGregorian y m 1)) == 7 ]

